### ISSUE TITLE
new openssh 1:9.0p1-1: our escapes are "broken" (not needed); empties out conflicting target file

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#6655]
TL;DR I thought I was going nuts since the test must have worked before, but while troubleshooting #6273 fails I ran into an odd failure due to target directory with extension `.c opy` becomes `.c\\\ opy`.  I have tested on master to the same extent. Then I checked openssh version, downgraded to `1:8.4p1-5` and the test passed!

<details>
<summary>Gory details</summary> 

```shell
(git)lena:~datalad/datalad-master[master]git
$> DATALAD_TESTS_SSH=1 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --pdb datalad/support/tests/test_sshconnector.py:test_ssh_copy
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_get_most_obscure_supported_nameg71d2wbi 
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_get_most_obscure_supported_name0fkb244d 
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_ilduxyc4 
datalad.support.tests.test_sshconnector.test_ssh_copy ... > /home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.9/site-packages/nose/tools/trivial.py(22)ok_()
-> raise AssertionError(msg)
(Pdb) up
> /home/yoh/proj/datalad/datalad-master/datalad/support/tests/test_sshconnector.py(235)test_ssh_copy()
-> ok_(isdir(targetdir))
(Pdb) p targetdir
'/home/yoh/.tmp/datalad_temp_test_ssh_copyxklo1b8i.c opy'
(Pdb) p isdir(targetdir)
False
*(Pdb) p op.lexists(targetdir)
False
(Pdb) 
[1]  + 84808 suspended  DATALAD_TESTS_SSH=1 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --pdb 
(dev3) 1 20590 ->148 [1].....................................:Tue 26 Apr 2022 06:07:54 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> ls -ld /home/yoh/.tmp/datalad_temp_test_ssh_copyxklo1b8i.c*
drwx------ 1 yoh yoh 200 Apr 26 18:07 /home/yoh/.tmp/datalad_temp_test_ssh_copyxklo1b8i.c\\\ opy/
(dev3) 1 20591 [1].....................................:Tue 26 Apr 2022 06:08:02 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> fg
[1]  + 84808 continued  DATALAD_TESTS_SSH=1 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --pdb 


False
*(Pdb) p isdir(targetdir.replace(' ', r'\ '))
True
(Pdb) 
[1]  + 84808 suspended  DATALAD_TESTS_SSH=1 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --pdb 
(dev3) 1 20592 ->148 [1].....................................:Tue 26 Apr 2022 06:08:39 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> ssh --version
unknown option -- -
usage: ssh [-46AaCfGgKkMNnqsTtVvXxYy] [-B bind_interface]
           [-b bind_address] [-c cipher_spec] [-D [bind_address:]port]
           [-E log_file] [-e escape_char] [-F configfile] [-I pkcs11]
           [-i identity_file] [-J [user@]host[:port]] [-L address]
           [-l login_name] [-m mac_spec] [-O ctl_cmd] [-o option] [-p port]
           [-Q query_option] [-R address] [-S ctl_path] [-W host:port]
           [-w local_tun[:remote_tun]] destination [command [argument ...]]
(dev3) 1 20593 ->255 [1].....................................:Tue 26 Apr 2022 06:08:41 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> acpolicy ssh-client
ssh-client:
  Installed: (none)
  Candidate: (none)
  Version table:
(dev3) 1 20594 [1].....................................:Tue 26 Apr 2022 06:08:48 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> acpolicy openssh-client
openssh-client:
  Installed: 1:9.0p1-1
  Candidate: 1:9.0p1-1
  Version table:
 *** 1:9.0p1-1 900
        900 http://deb.debian.org/debian bookworm/main amd64 Packages
        600 http://http.debian.net/debian sid/main amd64 Packages
        100 /var/lib/dpkg/status
     1:8.4p1-5 100
        100 http://deb.debian.org/debian bullseye/main amd64 Packages
(dev3) 1 20595 [1].....................................:Tue 26 Apr 2022 06:08:52 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> sudo apt-get install openssh-client=1:8.4p1-5
[sudo] password for yoh: 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following packages were automatically installed and are no longer required:
  android-libadb android-libart android-libboringssl android-libcrypto-utils android-libetc1 android-libext4-utils android-libnativebridge android-libnativeloader
  android-libsparse dexdump exfat-fuse gcc-10-base:i386 gcc-11-base:i386 gir1.2-dazzle-1.0 gir1.2-gnomebluetooth-1.0 gir1.2-gweather-3.0 golang-1.17 golang-1.17-doc golang-1.17-go
  golang-1.17-src gridengine-drmaa1.0 libabsl20200923 libcmark-gfm-extensions0 libcmark-gfm-extensions0.29.0.gfm.2 libcmark-gfm0 libcmark-gfm0.29.0.gfm.2 libdazzle-1.0-0
  libdazzle-common libfam0 libffms2-4 libfl2 libfwupdplugin1 libgavl1 libgeos3.10.1 libgpac10 libgweather-3-16 libgweather-common libhash-fieldhash-perl libhypre-2.22.1 libicu70
  libicu71 libindicator3-7 libjs-d3-format libjs-jquery-hotkeys libjs-jquery-isonscreen libjs-jquery-metadata libjs-jquery-tablesorter libjs-jquery-throttle-debounce libjsoncpp24
  libldap-2.4-2 libllvm12 libllvm12:i386 libmbedcrypto3 libmbedtls12 libmbedx509-0 libmoox-struct-perl libmozjs-78-0 libnetpbm10 libnode93 libobject-id-perl liboctave-dev libofa0
  libperl5.32 libpetsc-real3.15 libplacebo120 libplacebo157 libpoppler102 libproj22 libptscotch-6.1 libscalapack-openmpi2.1 libscotch-6.1 libsdbus-c++0 libsdl-ttf2.0-0 libslurm36
  libstd-rust-1.56 libtrilinos-amesos12 libtrilinos-aztecoo12 libtrilinos-epetra12 libtrilinos-epetraext12 libtrilinos-galeri12 libtrilinos-ifpack12 libtrilinos-kokkos12
  libtrilinos-ml12 libtrilinos-teuchos12 libtrilinos-trilinosss12 libtrilinos-triutils12 libtrilinos-zoltan12 libwebp6:i386 libwmf0.2-7 libyara8 linux-compiler-gcc-10-x86
  linux-headers-5.14.0-2-amd64 linux-headers-5.14.0-2-common linux-headers-5.15.0-2-amd64 linux-headers-5.15.0-2-common linux-image-5.14.0-2-amd64 linux-image-5.15.0-2-amd64
  linux-kbuild-5.14 mocha node-ansi-colors node-browser-stdout node-clean-yaml-object node-d3 node-d3-array node-d3-axis node-d3-brush node-d3-chord node-d3-collection
  node-d3-color node-d3-contour node-d3-dispatch node-d3-drag node-d3-dsv node-d3-ease node-d3-fetch node-d3-force node-d3-format node-d3-geo node-d3-hierarchy node-d3-interpolate
  node-d3-path node-d3-polygon node-d3-quadtree node-d3-queue node-d3-random node-d3-scale node-d3-scale-chromatic node-d3-selection node-d3-shape node-d3-time node-d3-time-format
  node-d3-timer node-d3-transition node-d3-voronoi node-d3-zoom node-dagre-d3-renderer node-dagre-layout node-execa node-graphlibrary node-he node-invert-kv node-lcid node-mem
  node-merge-stream node-mimic-fn node-os-locale node-postcss node-randombytes node-require-main-filename node-rw node-serialize-javascript node-strip-eof node-strip-json-comments
  node-tmatch node-which-module ocaml-base-nox perl-modules-5.32 python3-cov-core python3-crcmod python3-drmaa python3-mistune python3-pytest-cov python3-pytest-sugar
  python3-rdflib-jsonld python3-shellingham python3-termcolor python3-twisted-bin ruby2.7 ruby2.7-dev ruby2.7-doc x11proto-xinerama-dev
Use 'sudo apt autoremove' to remove them.
Suggested packages:
  libpam-ssh monkeysphere
The following packages will be REMOVED:
  openssh-server openssh-sftp-server
The following packages will be DOWNGRADED:
  openssh-client
0 upgraded, 0 newly installed, 1 downgraded, 2 to remove and 17 not upgraded.
Need to get 0 B/929 kB of archives.
After this operation, 3,403 kB disk space will be freed.
Do you want to continue? [Y/n] 
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
	modified:   ansible (modified content, untracked content)

no changes added to commit (use "git add" and/or "git commit -a")
warning: etckeeper failed to commit changes in /etc using git
(Reading database ... 1136573 files and directories currently installed.)
Removing openssh-server (1:9.0p1-1) ...
Removing openssh-sftp-server (1:9.0p1-1) ...
dpkg: warning: downgrading openssh-client from 1:9.0p1-1 to 1:8.4p1-5
(Reading database ... 1136546 files and directories currently installed.)
Preparing to unpack .../openssh-client_1%3a8.4p1-5_amd64.deb ...
Unpacking openssh-client (1:8.4p1-5) over (1:9.0p1-1) ...
Setting up openssh-client (1:8.4p1-5) ...
Processing triggers for man-db (2.10.2-1) ...
[master ec7affd] committing changes in /etc made by "apt-get install openssh-client=1:8.4p1-5"
 Author: Yaroslav Halchenko <debian@onerussian.com>
 10 files changed, 6 insertions(+), 5 deletions(-)
 delete mode 120000 runit/runsvdir/default/ssh
 delete mode 120000 sv/ssh/log/supervise
 delete mode 120000 sv/ssh/supervise
 create mode 120000 systemd/system/ssh.service
 create mode 120000 systemd/system/ssh.socket
Scanning processes...                                                                                                                                                                
Scanning candidates...                                                                                                                                                               
Scanning processor microcode...                                                                                                                                                      
Scanning linux images...                                                                                                                                                             

Running kernel seems to be up-to-date.

The processor microcode seems to be up-to-date.

Restarting services...
Service restarts being deferred:
 systemctl restart user@47521.service

No containers need to be restarted.

No user sessions are running outdated binaries.
(dev3) 1 20596 [1].....................................:Tue 26 Apr 2022 06:09:33 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> DATALAD_TESTS_SSH=1 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --pdb datalad/support/tests/test_sshconnector.py:test_ssh_copy
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_get_most_obscure_supported_name6tjo0y19 
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_get_most_obscure_supported_namey86mtei8 
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_04hzlbbm 
datalad.support.tests.test_sshconnector.test_ssh_copy ... ok
Versions: annexremote=1.5.0 boto=2.49.0 cmd:7z=16.02 cmd:annex=8.20211123 cmd:bundled-git=UNKNOWN cmd:git=2.35.1 cmd:system-git=2.35.1 cmd:system-ssh=8.4p1 exifread=2.3.2 humanize=3.14.0 iso8601=1.0.2 keyring=23.5.0 keyrings.alt=UNKNOWN msgpack=1.0.3 mutagen=1.45.1 platformdirs=2.5.0 requests=2.27.1
Obscure filename: str=b' |;&%b5{}\'"\xce\x94\xd0\x99\xd7\xa7\xd9\x85\xe0\xb9\x97\xe3\x81\x82 .datc ' repr=' |;&%b5{}\'"ΔЙקم๗あ .datc '
Encodings: default='utf-8' filesystem='utf-8' locale.prefered='UTF-8'
Environment: PATH='/home/yoh/proj/datalad/datalad-master/venvs/dev3/bin:/home/yoh/gocode/bin:/home/yoh/gocode/bin:/home/yoh/bin:/home/yoh/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/sbin:/usr/sbin:/usr/local/sbin' LANG='en_US.UTF-8' GIT_PAGER='less --no-init --quit-if-one-screen' GIT_CONFIG_PARAMETERS="'init.defaultBranch=dl-test-branch' 'clone.defaultRemoteName=dl-test-remote'" GIT_CONFIG_GLOBAL='/home/yoh/.tmp/datalad_temp_04hzlbbm/.gitconfig' GIT_ASKPASS='true'
[INFO   ] Keeping temp file: /home/yoh/.tmp/datalad_temp_04hzlbbm 
[INFO   ] Keeping temp file: /home/yoh/.tmp/httpserveq3foqoy5 

----------------------------------------------------------------------
Ran 1 test in 0.234s

OK
```
</details>

More specific demo:

<details>
<summary>local: OpenSSH_9.0p1 Debian-1, OpenSSL 1.1.1n  15 Mar 2022</summary> 

```shell
$> rm -rf /tmp/uc*; scp auth.c 'localhost:/tmp/uc opy'; ls -ld /tmp/uc*
auth.c                                                                                                                                             100%   28KB  19.9MB/s   00:00    
-rw------- 1 yoh yoh 28253 Apr 26 18:22 /tmp/uc\ opy

$> rm -rf /tmp/uc*; scp auth.c 'localhost:/tmp/uc\ opy'; ls -ld /tmp/uc*
auth.c                                                                                                                                             100%   28KB  17.4MB/s   00:00    
-rw------- 1 yoh yoh 28253 Apr 26 18:22 /tmp/uc\\\ opy

```
</details>

<details>
<summary>scping to smaug with OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1k  25 Mar 2021</summary> 

```shell
$> ssh -V
OpenSSH_9.0p1 Debian-1, OpenSSL 1.1.1n  15 Mar 2022

$> ssh smaug.datalad.org ssh -V
OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1k  25 Mar 2021

$> rm -rf /tmp/uc*; scp auth.c 'smaug:/tmp/uc\ opy'; ssh smaug 'ls -ld /tmp/uc*'
auth.c                                                                                                                                             100%   28KB 838.2KB/s   00:00    
-rw------- 1 yoh yoh 28253 Apr 26 18:25 /tmp/uc\ opy

$> rm -rf /tmp/uc*; scp auth.c 'smaug:/tmp/uc opy'; ssh smaug 'ls -ld /tmp/uc*' 
auth.c                                                                                                                                             100%   28KB 862.0KB/s   00:00    
-rw------- 1 yoh yoh 28253 Apr 26 18:26 /tmp/uc opy
-rw------- 1 yoh yoh 28253 Apr 26 18:25 /tmp/uc\ opy

```
</details>

<details>
<summary>when both client/server old locally: OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1n  15 Mar 2022 -- can't copy without escaping</summary> 

```shell

$> rm -rf /tmp/uc*; scp auth.c 'localhost:/tmp/uc\ opy'; ls -ld /tmp/uc*
auth.c                                                                                                                                                                             100%   28KB  23.5MB/s   00:00    
-rw------- 1 yoh yoh 28253 Apr 26 20:41 /tmp/uc\ opy

$> rm -rf /tmp/uc*; scp auth.c 'localhost:/tmp/uc opy'; ls -ld /tmp/uc* 
scp: ambiguous target
ls: cannot access '/tmp/uc*': No such file or directory

$> ssh -V
OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1n  15 Mar 2022
```
</details>
- so it seems to be largely server (not client) version dependent!  That would make it trickier to test/apply escaping dynamically if that is really needed
- but also it simply doesn't work without escaping if both are old
- [OpenSSH release notes](https://www.openssh.com/releasenotes.html) mentions "both the client and server are now using a  stricter configuration file parser. The new parser uses more  shell-like rules for quotes, space and escape characters." for [8.7](https://www.openssh.com/txt/release-8.7)/[8.7p1](https://www.openssh.com/txt/release-8.7)
- I do not yet quite grasp yet how it works "ok" (in that test passes) with 8.4p since it seems if `\` is specified -- it is preserved in the name.

### Another gotcha (of our test as well)

In [test_ssh_copy](https://github.com/datalad/datalad/blob/HEAD/datalad/support/tests/test_sshconnector.py#L211) if running against "localhost" (i.e. I have datalad-test to be 128.0.0.1 in my /etc/hosts) we pretty much copy that "obscure" file name over itself.

<details>
<summary>and surprise -- with old ssh it is ok -- it survives</summary> 

```shell
$> ssh -V; echo 123 >| 123 && scp 123 localhost:$PWD/123 && cat 123
OpenSSH_8.4p1 Debian-5, OpenSSL 1.1.1n  15 Mar 2022
123                                                                     100%    4    15.0KB/s   00:00    
123
```
</details>

<details>
<summary>while with new one -- you get it silently emptied out</summary> 

```shell
$> ssh -V; echo 123 >| 123 && scp 123 localhost:$PWD/123 && cat 123 && ls -l 123           
OpenSSH_9.0p1 Debian-1, OpenSSL 1.1.1n  15 Mar 2022
123                                                                       0%    0     0.0KB/s   --:-- ETA
-rw------- 1 yoh yoh 0 Apr 26 21:03 123
```
</details>

I don't think we should copy that poor file over itself, but also I think there is a bug in ssh :-/

**Comment by adswa** [comment]
I read trough the chain of issues and PRs connected to this issue, and I was wondering if #6826 addressed any aspect of this issue as well?
